File: projecteuler/p0001/multiples_of_3_or_5.go
package p0001

func bruteForce(n int) int {
	sum := 0
	for i := 0; i < n; i++ {
		if i%3 == 0 || i%5 == 0 {
			sum += i
		}
	}
	return sum
}

func linearSum(k int) int {
	return k * (k + 1) / 2
}

func closedForm(n int) int {
	m := n - 1
	return 3*linearSum(m/3) +
		5*linearSum(m/5) -
		15*linearSum(m/15)
}

func goroutineWithChannel(n int) int {
	divisors := make(chan int)
	go func() {
		defer close(divisors)
		for i := 0; i < n; i++ {
			if i%3 == 0 || i%5 == 0 {
				divisors <- i
			}
		}
	}()

	sum := 0
	for i := range divisors {
		sum += i
	}
	return sum
}

func bufferedChannel(n int) int {
	divisors := make(chan int, n)
	for i := 0; i < n; i++ {
		if i%3 == 0 || i%5 == 0 {
			divisors <- i
		}
	}
	close(divisors)

	sum := 0
	for i := range divisors {
		sum += i
	}
	return sum
}

File: projecteuler/p0001/multiples_of_3_or_5.ts
/**
 * Calculates the arithmetic sum of a sequence according to:
 * http://www.mathsisfun.com/algebra/sequences-sums-arithmetic.html
 */
const arithmetic_sum = function (a: number, d: number, n: number) {
  return (n / 2) * (2 * a + d * (n - 1));
};

/**
 * Calculates the sum of the divisors {b, c} in range (1, upper)
 */
export const sum_divisors = function (
  upper: number,
  div1: number,
  div2: number
) {
  // The range is exclusive, so subtract 1 to get the inclusive upperbound
  const upper_inclusive = upper - 1;
  // Need the combo of div1 * div2 to subtract divisor double-counts
  const c_div = div1 * div2;

  return (
    arithmetic_sum(div1, div1, Math.floor(upper_inclusive / div1)) +
    arithmetic_sum(div2, div2, Math.floor(upper_inclusive / div2)) -
    arithmetic_sum(c_div, c_div, Math.floor(upper_inclusive / c_div))
  );
};

export const brute_force = function (n: number) {
  let sum = 0;
  for (let i = 0; i < n; i++) {
    if (i % 3 === 0 || i % 5 === 0) {
      sum += i;
    }
  }
  return sum;
};
